src/test/java/org/prebid/server/auction/ExchangeServiceTest.java

```diff
@@ -3500,6 +3622,89 @@ public void shouldReturnBidResponseWithHooksVerboseTraceInfoWhenAuctionHappened(
+                .willReturn(Future.succeededFuture(givenSeatBid(emptyList())));
+
+        final BidRequest bidRequest = givenBidRequest(givenSingleImp(singletonMap("bidder", 2)));
+        final AuctionContext auctionContext = givenRequestContext(bidRequest).toBuilder()
+                .hookExecutionContext(HookExecutionContext.of(
+                        Endpoint.openrtb2_auction,
+                        stageOutcomes(givenAppliedToImpl(appliedToImplBuilder -> appliedToImplBuilder
+                                .impIds(asList("impId1", "impId2"))
+                                .response(true)
+                                .request(false)))))
+                .debugContext(DebugContext.of(false, TraceLevel.verbose))
+                .build();
+
+        // when
+        final AuctionContext result = exchangeService.holdAuction(auctionContext).result();
+
+        // then
+        final BidResponse bidResponse = result.getBidResponse();
+        assertThat(bidResponse.getExt().getPrebid().getModules().getTrace().getStages())
+                .anySatisfy(stage -> assertThat(stage.getOutcomes())
+                        .anySatisfy(outcome -> assertThat(outcome.getGroups())
+                                .anySatisfy(group -> assertThat(group.getInvocationResults())
+                                        .anySatisfy(hook -> {
+                                            assertThat(hook.getDebugMessages())
+                                                    .containsOnly("debug message 1-1 1", "debug message 1-1 2");
+                                            assertThat(hook.getAnalyticsTags()).isEqualTo(
+                                                    ExtModulesTraceAnalyticsTags.of(singletonList(
+                                                            ExtModulesTraceAnalyticsActivity.of(
+                                                                    "some-activity",
+                                                                    "success",
+                                                                    singletonList(ExtModulesTraceAnalyticsResult.of(
+                                                                            "success",
+                                                                            mapper.createObjectNode(),
+                                                                            ExtModulesTraceAnalyticsAppliedTo.builder()
+                                                                                    .impIds(asList("impId1", "impId2"))
+                                                                                    .response(true)
+                                                                                    .request(null)
+                                                                                    .build()))))));
+                                        }))));
```

Comment:
Pls, rewrite this test to be more linear, not nested. For now it's barely readable.